### PR TITLE
#181339527 User Role Settings

### PR DESCRIPTION
**Pull request by Tsa2341**
### What does this PR do?

this PR implements roles of user

### Description of Task to be completed?

- Give a user a role on the register
- Give a SUPER_ADMIN ability to change the role of any user
- Protect the route if not the SUPER_ADMIN logged in

### How should this be manually tested?

- super_admin credentials
         - email: SUPER_ADMIN@gmail.com
         - password: SUPER_ADMIN2gmail

- remotely: go to this link: https://staging-bare-feature-18-rszml6.herokuapp.com/api-docs/#/
- locally: clone this repo
- checkout to feature/#23181339527-user-role-settings branch and run `yarn run start`
- use postman and login and get the token
         - POST    **http://localhost:3000/api/v1/user/login**
- use postman to test this API with the token in headers as a Bearer : 
          - PATCH **http://localhost:3000/api/v1/users/assignRole**
          -  GET**http://localhost:3000/api/v1/users/getRoles**

### What are the relevant pivotal tracker stories?

[#181339527](https://www.pivotaltracker.com/story/show/181339527)

### Screenshots (if appropriate)

Screenshot (20)
Screenshot (21)

### CODE REVIEW

**Review by Code-250 — APPROVED** [review]

**Review by tuyizerejean — APPROVED** [review]

File: src/utils/roles.utils.js [hunk 1]
```diff
@@ -1,8 +1,3 @@
-const roles = {
-	SUPER_ADMIN: '94339a6a-2fc1-4bfc-814c-616b72c87306',
-	TRAVEL_ADMIN: 'f24659cc-e9bd-43ee-b97a-294b3c154a78',
-	MANAGER: '69263f66-8890-419a-97e7-b9f2f1e5e578',
-	REQUESTER: '0a422b6d-d6bc-4373-ba91-c51268323fc1',
-};
+const roles = ['SUPER_ADMIN','TRAVEL_ADMIN','MANAGER','REQUESTER',];
 
 module.exports = roles;
```
Comment (ncutixavier) [inline src/utils/roles.utils.js:3]: Use ES6

File: src/helpers/user.helpers.js [hunk 2]
```diff
@@ -14,13 +15,13 @@ function comparePassword(plainPassword, hashedPassword) {
...
-// 	const verify = jwt.verify(token, process.env.SECRETE)
-// 	return verify
-// }
+function decodeToken(token) {
+  const verify = jwt.verify(token, process.env.SECRETE);
+  return verify;
+}
 
-module.exports = { hashPassword, comparePassword, generateToken };
+module.exports = { hashPassword, comparePassword, generateToken, decodeToken };
```
Comment (ncutixavier) [inline src/helpers/user.helpers.js:27]: Use ES6

File: src/controllers/user.controller.js [hunk 3]
```diff
@@ -27,22 +28,26 @@ export default class UserController {
...
+      if (validation) {
+        const token = await generateToken({ email: user.email }, '1d');
+        return res
+          .status(201)
+          .header('authenticate', token)
+          .json({ message: 'Logged in successfully', token });
       }
+      return res.status(400).json({ message: 'Invalid password' });
     } catch (error) {
       return res.status(404).json({
```
Comment (ncutixavier) [inline src/controllers/user.controller.js:47]: Change status code

File: src/controllers/role.controller.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      if (!user) {
+        return res.status(404).json({
+          message: `User with email ${email} doesn't exist`
+        });
+      }
+      if (user.role_id === roleId) {
+        return res.status(409).json({
+          message: `User already have ${role} role`
+        });
+      }
```
Comment (ncutixavier) [inline src/controllers/role.controller.js]: This should be implemented under `role.middleware.js`

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/utils/roles.utils.js (hunk 1 above)
Comment (Tsa2341) [reply src/utils/roles.utils.js:3]: The roles and the hashPassword functions are being used in models and seeders files which when running with sequelize they do not use babel so thereby requiring to use es5 files even in import

File: src/helpers/user.helpers.js (hunk 2 above)
Comment (Tsa2341) [reply src/helpers/user.helpers.js:27]: The roles and the hashPassword functions are being used in models and seeders files which when running with sequelize they do not use babel so thereby requiring to use es5 files even in import

**Review by ncutixavier — APPROVED** [review]

**Pull request merged**